hoomd/conftest.py

```diff
@@ -271,3 +324,128 @@ def operation_pickling_check(instance, sim):
     sim.operations += instance
     sim.run(0)
     pickling_check(instance)
+
+
+class BlockAverage:
+    """Block average method for estimating standard deviation of the mean.
+
+    Args:
+        data: List of values
+    """
+
+    def __init__(self, data):
+        # round down to the nearest power of 2
+        N = 2**int(math.floor(math.log(len(data)) / math.log(2)))
+        if N != len(data):
+            warnings.warn(
+                "Ignoring some data. Data array should be a power of 2.")
+
+        block_sizes = []
+        block_mean = []
+        block_variance = []
+
+        # take means and averages of blocks, growing blocks by factors of 2
+        block_size = 1
+        while block_size <= N // 8:
+            num_blocks = N // block_size
+            block_data = numpy.zeros(num_blocks)
+
+            for i in range(0, num_blocks):
+                start = i * block_size
+                end = start + block_size
+                block_data[i] = numpy.mean(data[start:end])
+
+            block_mean.append(numpy.mean(block_data))
+            block_variance.append(numpy.var(block_data) / (num_blocks - 1))
+
+            block_sizes.append(block_size)
+            block_size *= 2
+
+        self._block_mean = numpy.array(block_mean)
+        self._block_variance = numpy.array(block_variance)
+        self._block_sizes = numpy.array(block_sizes)
+        self.data = numpy.array(data)
+
+        # check for a plateau in the relative error before the last data point
+        block_relative_error = numpy.sqrt(self._block_variance) / numpy.fabs(
+            self._block_mean)
+        relative_error_derivative = (
+            block_relative_error[1:] - block_relative_error[:-1]) / (
+                self._block_sizes[1:] - self._block_sizes[:-1])
+        if numpy.all(relative_error_derivative > 0):
```

Comment:
The goal of this is to check if there is a plateau in the error. The errors tend to grow very quickly for the first few blockings, then flatten out - like [1,2,10,20,21,19,20]. The >0 check is meant to be a simple check to see if the samples are still in the growing stage. If they are, issue a warning that the simulation likely needs to be run longer.

Reply:
What I was thinking about was something like `[1, 2, 10, 20, 21, 21.1, 21.100001]`. I know there will be some noise that will in most cases prevent such a one-sided approach, but this relies on a larger size reducing the error at some point which in a true plateau doesn't happen, unless I am misunderstanding something.

Reply:
Ultimately, this isn't used as a hard check in the validation, just a soft warning to note that maybe the reason the test is failing is because the error analysis did not plateau. In developing the validation test and determining how long to run it, the developer should print the error vs block size and inspect how it behaves by eye.

I'd prefer to run 4 independent runs and to the standard deviation of the mans, but I don't think we can afford that much time.